### PR TITLE
Hotspot/Large pages: FreeBSD support proposal.

### PR DESCRIPTION
**Pull request by devnexen**
Enabling UseLargePages flag but only via FreeBSD's super page's feature,
 aligned to the arch page size.
It exists since ~2008, so it is safe here and can be enabled/disabled
at boot time only so we check it regardless.

### CODE REVIEW

**Comment by devnexen** [comment]
Do not know if it s the proper repository for new features rather than openjdk12 ? Tested with couple of simple apps.

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (bsdkurt) [inline src/hotspot/os/bsd/os_bsd.cpp:2331]: This value is exposed to shared hotspot code via os::large_page_size(). Please use `#if defined(__FreeBSD__)` for the non-zero values.

**Review by bsdkurt — CHANGES_REQUESTED** [review]

**Comment by bsdkurt** [comment]
Targeting jdk11 with new features is fine with me. @battleblow should weigh in on this as well.

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (devnexen) [reply src/hotspot/os/bsd/os_bsd.cpp:2331]: Ah correct, OpenBSD has no equivalent feature, not sure about NetBSD but do not think neither. MacOS would possibly be the only "bsd" which would share few values of these.

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (bsdkurt) [inline src/hotspot/os/bsd/os_bsd.cpp:2344]: Looking at other OSs implementation they short circuit exit if UseLargePages is not true here.
`if (!UseLargePages) return;`

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (bsdkurt) [inline src/hotspot/os/bsd/os_bsd.cpp]: If UseLargePages was true but can't be used, other OSs disable UseLargePages. Please replace _large_page_size = 0 with `UseLargePages = false`. Without this change I believe the assert at line 2357 will be tripped if the system is not configured for large pages.

**Review by bsdkurt — CHANGES_REQUESTED** [review]
In addition to the minor changes below, I believe that changes are needed with `os::can_commit_large_page_memory() and `os::can_execute_large_page_memory()`.

Currently they are using `UseHugeTLBFS` but I believe that is incorrect. It looks like `UseHugeTLBFS` is linux only and was accidentally ported over to bsd. I believe this flag should be removed from `src/hotspot/os/bsd/globals_bsd.hpp` and `os::can_commit_large_page_memory()` and `os::can_execute_large_page_memory()` be changed to match your implementation. Since you have provided an implementation for `os::reserve_memory_special` and it supports PROT_EXEC I believe the following is correct:
```
bool os::can_commit_large_page_memory() {
  return true;
}

bool os::can_execute_large_page_memory() {
  return true;
}
```
Edit: fix return from os::can_commit_large_page_memory

**Comment by devnexen** [comment]
> In addition to the minor changes below, I believe that changes are needed with `os::can_commit_large_page_memory() and `os::can_execute_large_page_memory()`.
> 
> Currently they are using `UseHugeTLBFS` but I believe that is incorrect. It looks like `UseHugeTLBFS` is linux only and was accidentally ported over to bsd.

Yes you re right it is a pure Linux-ism, UseLargePages alone is relevant here.

**Comment by bsdkurt** [comment]
Hmm, after reading src/hotspot/share/memory/virtualspace.cpp I think os::reserve_memory_special() is only called if os::can_commit_large_page_memory() returns false. so I may have inverted the logic in my initial recommendation.

**Comment by devnexen** [comment]
No problem I appreciate your comments :-)

**Review by bsdkurt — APPROVED** [review]

**Comment by bsdkurt** [comment]
From a code review standpoint I don't see anything further. Thanks for working on this.

I'd suggest running the hotspot tests under test/hotspot/jtreg on your branch. There are specific tests for UseLargePages in there that should be helpful to test it is working well. Also I think Greg will want to review this before merging so I added him as a reviewer.

**Comment by devnexen** [comment]
Tests as : 
- `test/hotspot/jtreg/gc/arguments/TestAlignmentToUseLargePages.java`
- `test/hotspot/jtreg/gc/TestSystemGC.java`
- `test/hotspot/jtreg/runtime/memory/LargePages/TestLargePagesFlags.java`
- `test/hotspot/jtreg/vmTestbase/gc/memory/LargePagesTest/LargePageSizeInBytes.java`
- `test/hotspot/jtreg/vmTestbase/gc/memory/LargePagesTest/LargePagesTest.java`

passed.

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (battleblow) [inline src/hotspot/os/bsd/os_bsd.cpp:2332]: My comment applies to the whole block here.  (I obviously need to figure out to do that within the Github review system, but since I haven't I am stating it instead)

What assignment does this code make on FreeBSD/i386?  That appears to fall into none of the conditionals here and seems like it would yield a compile error.

File: src/hotspot/share/runtime/arguments.cpp
Comment (battleblow) [inline src/hotspot/share/runtime/arguments.cpp]: Nit: Needs a space between )) and ||

File: src/hotspot/share/runtime/arguments.cpp
Comment (battleblow) [inline src/hotspot/share/runtime/arguments.cpp]: The logic here looks incorrect to me.  Compare the change you made on line 3852.  Previously the conditional here was the logical inverse of the conditional there.  It now isn't.  The simplest way to keep the two conditionals in sync may be to simply make one the direct inverse of the other using !() around it.

My understanding here is that the intent of the change is that UseLargePages now gets set to true on FreeBSD, but not on other BSD.

**Review by battleblow — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/arguments.cpp
Comment (devnexen) [reply src/hotspot/share/runtime/arguments.cpp]: good catch indeed.

File: src/hotspot/os/bsd/os_bsd.cpp
Comment (devnexen) [reply src/hotspot/os/bsd/os_bsd.cpp:2332]: Nope sorry my bad  I naively thought AMD64 covered it  ...

**Review by battleblow — COMMENTED** [review]
This looks good to me.  Have you run the test cases since the most recent changes?

**Comment by devnexen** [comment]
Yes just reran them.

**Pull request merged**
